### PR TITLE
librbd: clone copy-on-write operations should preserve sparseness

### PR DESCRIPTION
**Pull request by trociny**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman thanks! updated

File: src/librbd/image/RefreshRequest.cc [hunk 1]
```diff
@@ -1320,6 +1320,13 @@ void RefreshRequest<I>::apply() {
       m_image_ctx.parent_md = m_parent_md;
       m_image_ctx.migration_info = {};
     }
+
+    librados::Rados rados(m_image_ctx.md_ctx);
+    int8_t require_osd_release;
+    int r = rados.get_min_compatible_osd(&require_osd_release);
+    if (r == 0 && require_osd_release >= CEPH_RELEASE_OCTOPUS) {
```
Comment (dillaman) [inline src/librbd/image/RefreshRequest.cc]: Need to rebase and tweak

File: src/librbd/io/CopyupRequest.h [hunk 2]
```diff
@@ -87,7 +88,9 @@ class CopyupRequest {
   bool m_flatten = false;
   bool m_copyup_required = true;
   bool m_copyup_is_zero = true;
+  bool m_try_sparse_copyup =true;
```
[the remark this answers is not in the document]
Comment (dillaman) [reply src/librbd/io/CopyupRequest.h]: Nit: not need anymore

[the remark this answers is not in the document]
Comment (trociny) [reply src/librbd/io/CopyupRequest.h]: Thanks, done.

File: src/librbd/image/RefreshRequest.cc (hunk 1 above)
Comment (trociny) [reply src/librbd/image/RefreshRequest.cc]: @dillaman Not sure what you mean here. I rebased but there were no any confilicts, there seems nothing to tweak.

Comment (dillaman) [reply src/librbd/image/RefreshRequest.cc]: Never mind, I was thinking it might have been broken by PR 27855, but it wasn't.

**Comment by trociny** [comment]
@dillaman I had to tweak `TestInternal.SnapshotCopyup`: the previous version failed for `rbd_cache_policy=writeback`, due to reading from the writeback cache lost sparseness.

**Comment by dillaman** [comment]
retest this please

**Comment by dillaman** [comment]
retest this please

**Pull request merged**